### DISCUSSION TITLE
How to do custom title bar on Windows 10 using the GUI DSL?

### DISCUSSION THREAD

**Discussion by dumblob** [red/red#4972] [Q&A] [answered]
There are hard problems with custom title bar discussed in https://kubyshkin.name/posts/win32-window-custom-title-bar-caption/ along with a working solution to those.

Currently Red doesn't implement this which is a pity as people expect the Red "hackability" to extend over pretty much "anything".

Feel free to use these "hacks" to allow full customizability of the window (incl. title bar). This goes hand in hand with the trend "the app itself should draw everything itself" on other platforms (think Wayland, Mir, etc.).

**Comment by greggirwin** [comment]
Good article. Thanks for posting it.

> people expect the Red "hackability" to extend over pretty much "anything"

I don't know who you're talking to, but is that a reasonable expectation? :^) In any case, if it's done with APIs, you can probably do it with R/S (until we have FFI in Red). We can't support everything out of the box, and if this feature is product critical, it can be done. But there's also something to be said for following OS standards.

**Reply by dumblob** [reply]
> I don't know who you're talking to, but is that a reasonable expectation? :^)

IDK. It seems to stem from the API architecture.

>But there's also something to be said for following OS standards.

That's a hard question. But if M$ Office does that, M$ Edge does that, Google Chrome does that, etc. - then I'd call it an "OS standard". YMAMV though as always.

**Comment by hiiamboris** [comment, marked as answer]
Nothing stops you from removing OS title bar and drawing it yourself:
<img src=https://i.gyazo.com/6467bfe7e6ea4be04162400c62dc7de3.png width=200></img>
code: `view/flags/tight [base 300x20 draw [line-width 0 fill-pen linear magenta cyan box 0x0 300x20 text 1x1 "My custom title bar"] base 20x20 cyan draw [line 3x3 16x16 line 3x16 16x3] on-down [unview] return base white 320x300 "workspace"] 'no-title`

It's just extra work to reinvent the wheel (e.g. Alt+Space menu on Windows), and there's a possibility that other programs are using your title bar to place some controls.

**Reply by greggirwin** [reply]
Fantastic example @hiiamboris. Should be in a wiki, and just need to add window dragging from the title area for it to be complete.

**Reply by hiiamboris** [reply]
With window dragging: `p: none view/flags/tight [base 300x20 on-up [p: none] on-down [p: event/offset] all-over on-over [if p [fp: face/parent fp/offset: fp/offset + event/offset - p]] draw [line-width 0 fill-pen linear magenta cyan box 0x0 300x20 text 1x1 "My custom title bar"] base 20x20 cyan draw [line 3x3 16x16 line 3x16 16x3] on-down [unview] return base white 320x300 "workspace"] 'no-title`

**Reply by greggirwin** [reply]
https://github.com/red/red/wiki/%5BHOWTO%5D-VID-and-View-(GUI)-hints-and-tips#how-to-create-a-custom-title-bar-for-a-window
